lms/djangoapps/grades/models.py

```diff
@@ -0,0 +1,253 @@
+
+    def __unicode__(self):
+        """
+        Returns a string representation of this model.
+        """
+        return "PersistentSubsectionGrade for user {} in subsection {}".format(self.user_id, self.usage_key)
+
+    @classmethod
+    def save_grade(cls, **kwargs):
+        """
+        Wrapper for create_grade or update_grade, depending on which applies.
+        Takes the same arguments as both of thsoe methods.
+        """
+        try:
+            cls.update(**kwargs)
+        except cls.DoesNotExist:
+            cls.create(**kwargs)
+
+    @classmethod
+    def create(
+        cls,
+        user_id,
+        usage_key,
+        course_version,
+        subtree_edited_date,
+        earned_all,
+        possible_all,
+        earned_graded,
+        possible_graded,
+        visible_blocks,
+    ):
+        """
+        Instantiates a new model instance.
+
+        Example arguments:
+            user_id: "student12345"
+            usage_key: BlockUsageLocator object
+            course_version: "deadbeef"
+            subtree_edited_date: "2016-08-01 18:53:24.354741"
+            earned_all: 6
+            possible_all: 12
+            earned_graded: 6
+            possible_graded: 8
+            visible_blocks: [<list of BlockRecord objects>]
+        """
+        visible_blocks_model = VisibleBlocks.create(blocks=visible_blocks)
+
+        return cls.objects.create(
+            user_id=user_id,
+            course_id=usage_key.course_key,
+            usage_key=usage_key,
+            course_version=course_version,
+            subtree_edited_date=subtree_edited_date,
+            earned_all=earned_all,
+            possible_all=possible_all,
+            earned_graded=earned_graded,
+            possible_graded=possible_graded,
+            visible_blocks=visible_blocks_model,
+        )
+        return model
+
+    @classmethod
+    def read(cls, user_id, usage_key):
+        """
+        Reads a grade from database
+
+        Arguments:
+            user_id: The user associated with the desired grade
+            usage_key: The location of the subsection associated with the desired grade
+
+        Raises PersistentSubsectionGrade.DoesNotExist if applicable
+        """
+        return cls.objects.get(
+            user_id=user_id,
+            usage_key=usage_key,
+        )
+
+    @classmethod
+    def update(
+        cls,
+        user_id,
+        usage_key,
+        course_version,
+        subtree_edited_date,
+        earned_all,
+        possible_all,
+        earned_graded,
+        possible_graded,
+        visible_blocks,
+    ):
+        """
+        Updates a previously existing grade.
+
+        Requires all the arguments listed in docstring for create_grade
+        """
+        grade = cls.objects.get(
+            user_id=user_id,
+            usage_key=usage_key,
+        )
+
+        visible_blocks_model = VisibleBlocks.create(blocks=visible_blocks)
```

Comment:
It is very likely that we'll run into race conditions here. Please handle the case where you get an integrity error from multiple processes trying to insert the same blocks data at the same time. To see an example of this in our code, please look for `IntegrityError` in `course_overviews/models.py`